File: Two Pointers/345. Reverse Vowels of a String/345. Reverse Vowels of a String.js
/**
 * @param {string} s
 * @return {string}
 */
var reverseVowels = function (s) {
  const vowels = new Set("aeiouAEIOU");
  let l = 0;
  let r = s.length - 1;
  s = s.split("");

  while (l < r) {
    if (!vowels.has(s[l])) {
      l++;
      continue;
    }

    if (!vowels.has(s[r])) {
      r--;
      continue;
    }

    [s[l], s[r]] = [s[r], s[l]];
    l++;
    r--;
  }

  return s.join("");
};

// Time: O(n)
// Space: O(n)

File: Two Pointers/345. Reverse Vowels of a String/345. Reverse Vowels of a String.py
class Solution:
    def reverseVowels(self, s: str) -> str:
        vowels = set("aeiouAEIOU")
        s = list(s)
        l, r = 0, len(s) - 1

        while l < r:
            if s[l] not in vowels:
                l += 1
                continue
            
            if s[r] not in vowels:
                r -= 1
                continue

            s[l], s[r] = s[r], s[l]
            l += 1
            r -= 1
        
        return "".join(s)

# Time: O(n)
# Space: O(n)